### PR TITLE
ADDED: jax.lax.max function under frontend operators

### PR DESCRIPTION
**Pull request by SalihTalha**
jax.lax.max function under frontend operators, and it's test code. 
Issue: #2456

### CODE REVIEW

**Comment by SalihTalha** [comment]
Thank you very much for your review @VedPatwardhan . I have fixed the maximum part (I guess :D).

About the tests:
I have tried the function but I am keep getting error. I have not really figured out how dtype_and_values function works.

I have tried this to get one array with shape (2,1) and get 2 numbers out of it:
```py
    dtype_and_x=helpers.dtype_and_values(
        available_dtypes=tuple(
            set(ivy_np.valid_float_dtypes).intersection(set(ivy_jax.valid_float_dtypes))
        ),
        num_arrays=1,
        shared_dtype=True,
        shape=(2, 1)
    ),
```

and also this one to get x and y separately, then split their dtype and values.

```py
    dtype_and_x=helpers.dtype_and_values(
        available_dtypes=tuple(
            set(ivy_np.valid_float_dtypes).intersection(set(ivy_jax.valid_float_dtypes))
        ),
        num_arrays=2,
        shared_dtype=True,
    ),
    dtype_and_y=helpers.dtype_and_values(
        available_dtypes=tuple(
            set(ivy_np.valid_float_dtypes).intersection(set(ivy_jax.valid_float_dtypes))
        ),
        num_arrays=2,
        shared_dtype=True,
    ),
```

But as I debugged the code, I found that dtype_and_values function does not return and just gives error. How should I continue?

Thanks.

**Comment by VedPatwardhan** [comment]
Hi @SalihTalha:
1. You should update the way you're calling `dtype_and_values` in the tests, as we're comparing 2 arrays, you should pass `num_arrays=2` to it which would then draw data in the format `[dtype1, dtype2], [x1, x2]`, which you should then unpack and pass to the function, very similar to the `jax.lax.add` test, at the top of this file.
2. You should sample out the bools, i.e. `as_variable` and `native_array` using the `array_bools` helper by passing `num_array` to 2 as in your case.
3. All these argument names should also match the ones in the function.
4. When you call the `test_frontend_function` helper, you should pass both `x` and `y` as numpy arrays, as they provide a flexible way to convert to arrays in other types of backends, you can again refer the way the `jax.lax.add` function tests are written at the top of the file.
5. You should add the unsupported datatypes attribute in the `jax/operators.py` file for the `max` function exactly how it's done with other functions in that file.

Doing this should get the frontend tests to pass. Let me know if you have any other questions. Thanks 🙂

**Comment by SalihTalha** [comment]
Hey @VedPatwardhan,

I have tried that way when I first started but I come across some errors which I don't understand. But when I try it with --backend numpy it works now. Probably because my ivy environment does not have tensorflow & pytorch and that gives error. Can you check the code again please :) 

Little question: Why req.txt is so small, because even the test framework hypothesis is not in the requirements?

**Comment by VedPatwardhan** [comment]
> Hey @VedPatwardhan,
> 
> I have tried that way when I first started but I come across some errors which I don't understand. But when I try it with --backend numpy it works now. Probably because my ivy environment does not have tensorflow & pytorch and that gives error. Can you check the code again please :)
> 
> Little question: Why req.txt is so short, because even the test framework hypothesis is not in the requirements?

Actually, I have tested locally that the changes I'm suggesting would get the tests to pass, please share the details of the exact error message you're receiving, which could be helpful for understanding the problem.

Along with `requirements.txt`, `optional.txt` also has some more libraries listed in it, which includes `hypothesis`, so if you are working without docker, you should also install the libraries in `optional.txt`.

One last thing, please request my review through the right pane when you need my help, because otherwise I don't get a notification regarding the same. 

Let me know if you have any questions. Thanks🙂

**Comment by SalihTalha** [comment]
Hello again @VedPatwardhan , I've noted your tips.

About the test:
I first tried to install optional.txt packages and I get version incompatibility errors since I got no cuda installed currently on this machine. Then I tried to run it with Docker, followed the steps on ivy's website and it has no error messages currently.

I am very excited to continue my contributions, since I finally got the setup and settings done :D

Thank you so much for your help.

**Review by VedPatwardhan — COMMENTED** [review]
Great that you got everything setup! Feel free to re-request my review when you need my help 🙂

**Comment by VedPatwardhan** [comment]
Hey @SalihTalha, were you able to fix the issues? Let me know if you need my help by re-requesting my review via the pane at the right. Thanks 🙂

**Comment by SalihTalha** [comment]
Hey @VedPatwardhan , I don't know why *"test-frontend-numpy / run-nightly-tests"* gives error. It seems like it's because of some other functions. I can not relate it to my part of code. Can you help me?

**Review by VedPatwardhan — CHANGES_REQUESTED** [review]
Actually, I can see that you've added the function, but it seems like you've probably reverted the tests you had written. As there are no tests added for your function, none of the failures can be attributed to your task. Please add the test for your function which you had done a few commits ago and re-request my review, after which we'll look into getting the tests passing for your function if they fail.
Let me know if you have any questions. Thanks 🙂

**Comment by SalihTalha** [comment]
@VedPatwardhan 
Sorry for the delay :) I added the test codes again. I believe it is ready to merge.

**Comment by VedPatwardhan** [comment]
The tests are passing locally on my machine for your function, let's wait for the results in the CI and based on that we could merge your PR soon!

File: ivy_tests/test_ivy/test_frontends/test_jax/test_jax_lax_operators.py
Comment (VedPatwardhan) [inline ivy_tests/test_ivy/test_frontends/test_jax/test_jax_lax_operators.py:79]: Please add a blank line here

**Review by VedPatwardhan — CHANGES_REQUESTED** [review]
Hi @SalihTalha, the frontend tests are passing for your function, there's just one final change to fix the lint failure, after which the PR can be merged 🙂

**Comment by SalihTalha** [comment]
Added the newline to fix the lint failure.

**Review by VedPatwardhan — CHANGES_REQUESTED** [review]
Hi @SalihTalha, thanks for making the changes. There are a couple more changes to be made to your PR as a result of some updates on master, firstly, the `fn_name` argument passed to the `test_frontend_function` will now be renamed to `fn_tree`. Secondly, there have been some issues because of a commit yesterday on master, which are now resolved, so please pull the latest code on master.
Once you're done, feel free to re-request my review, if the tests pass, we can merge your PR.
Let me know if you have any questions. Thanks 🙂

**Comment by SalihTalha** [comment]
Hello @VedPatwardhan ,

I have made the change you mentioned. I got passed the tests on my local computer. I believe it is ready to merge.

**Review by VedPatwardhan — APPROVED** [review]
lgtm! Thanks for contributing 🙂

**Pull request merged**
